### ISSUE TITLE
`readonly` fields in a mapping-constructor 

### ISSUE THREAD

**Issue by MaryamZi** [ballerina-platform/ballerina-spec#553]
**Description:**

Few more points to clarify regarding $title.

**Case I**
When a mapping-constructor-expression has a `readonly` field, should the effective type of the field be narrowed down based on the actual expression specified as the value of the field?

For example, for

```ballerina
record {|
    string|int i;
|} rec = {
    readonly i: "hello"
};
```

should the type of the value constructed be

A)
```ballerina
record {|
    readonly string i; // just `string`
|}
```

or 
B)
```ballerina
record {|
    readonly string|int i; // still `string|int`
|}
```

I guess it can be A) since the field cannot be updated?

**Case II**

If a mapping-constructor-expression has a `readonly` field where the contextually expected type does not have an individual-type-descriptor for the particular field (i.e., a field of a map or a field that belongs to the rest-descriptor of a record), what would the resultant inherent type be?

Will the resultant type be a record type with individual-type-descriptors for such fields?

e.g.,

**Map**

For

```ballerina
map<string|int|boolean> mp = {
    i: "hello", 
    j: 1, 
    readonly k: true, 
    l: false
};
```

the resultant type

```ballerina
record {|
    readonly string|int|boolean k; // or `readonly boolean k` depending on Case I
    string|int|boolean...;
|}
```

**Record**
For

```ballerina
record {|
    string i;
    int j;
    boolean...;
|} rec = {
    i: "hello", 
    j: 1, 
    readonly "k": true, 
    "l": false
};
```

the resultant type

```ballerina
record {|
    string i;
    int j;
    readonly boolean k;
    boolean...;
|}
```

**Related Issues:**
https://github.com/ballerina-platform/ballerina-spec/issues/493

**Comment by MaryamZi** [comment]
Follow-up question related to Case II.

If the resultant type is a record type with individual-type-descriptors for `readonly` fields and a rest-descriptor for other fields, 
`map<string> mp = {readonly a: "hello", b: "ballerina"};` results in the following record type.

```ballerina
record {|
    readonly string a;
    string...;
|}
```

But since there is no field named `b`, using `{readonly a: "hello", b: "ballerina"}` with such a type is not usually allowed and we would have to use a string literal (`"b"`) as the key.

Should we 

1) just allow this specifically when a type is inferred due to the presence of `readonly` fields 

2) if there are `readonly` fields, mandate non-readonly fields to use string literals, if the expected type is a map?

`map<string> mp = {readonly a: "hello", "b": "ballerina"};`

3) or should the applicable type have individual-type-descriptors for the non-readonly fields too?

```ballerina
record {|
    readonly string a;
    string b;
    string...;
|}
```

**Comment by jclark** [comment]
When a mapping is entirely readonly, then the inherent type is a singleton: it contains a single shape that is the shape of the value. So when a single field of a mapping is readonly, the consistent rule would be that the inherent type of that field is a singleton.

So for case I, the answer is that the type would be `"hello"`.

I am not sure what you mean by _effective_ type and _applicable_ type.

Not sure if this answers your question.

**Comment by MaryamZi** [comment]
> When a mapping is entirely readonly, then the inherent type is a singleton: it contains a single shape that is the shape of the value. So when a single field of a mapping is readonly, the consistent rule would be that the inherent type of that field is a singleton.
> 
> So for case I, the inherent type of the constructed record would specify that the `i` field has type `"hello"` (i.e. a singleton type).
> 
Ack, yes, with literals we can use a singleton. And similarly if a non-literal expression is used (`i: getString()`) I guess we use `string` and not `string|int`.

> I am not sure what you mean by _effective_ type and _applicable_ type.

Sorry, let me rephrase. 

For something like

```ballerina
string s1 = "hello";
int i1 = 1;
boolean b1 = true;
boolean b2 = false;

map<string|int|boolean> mp = {
    i: s1, 
    j: i1, 
    readonly k: b1, 
    l: b2
};
```

since there are `readonly` fields, I believe the inherent type of the value will not be the same as the contextually expected type, as discussed in https://github.com/ballerina-platform/ballerina-spec/issues/493#issuecomment-615095572?

My question is regarding this actual inherent type.

Is it 

```ballerina
record {|
    readonly boolean k;
    string|int|boolean...;
|}
```

Similarly, for a record with `readonly` fields belonging to the rest-descriptor of the applicable contextually expected type

```ballerina
string s1 = "hello";
int i1 = 1;
boolean b1 = true;
boolean b2 = false;

record {|
    string i;
    int j;
    boolean...;
|} rec = {
    i: s1, 
    j: i1, 
    readonly "k": b1, 
    "l": b2
};
```

is it 

```ballerina
record {|
    string i;
    int j;
    readonly boolean k;
    boolean...;
|}
```

Basically, is it correct to assume `readonly` fields in a mapping-constructor-expression result in the inherent type of the created value to have individual-type-descriptors for such fields, even though the applicable contextually expected type doesn't have individual-type-descriptors for the same fields?

**Comment by jclark** [comment]
The concept of inherent type is an undesirable thing that we have only  in order to be able to handle mutation. If we did not have mutation, everything would be much simpler: we would just say a type is a set of values, and a value belongs to the type if it is a member of that set of values.  When dealing with immutable things we want to revert back to that simpler model, and the way we achieve that is to say that the inherent type is a singleton.

So when you have a mapping with a readonly field, the inherent type for that mapping should make the type for that field be a singleton. This is independent of what the contextually expected type is or of what type descriptor syntax happens to allow.

This means the answer to your question is "yes", although I wouldn't think of this in terms of creating individual-type-descriptors for the readonly fields.

**Comment by MaryamZi** [comment]
This is clear now, thank you.
